File: src/main/java/pw/mihou/amelia/clients/listeners/impl/FeedUpdater.java
package pw.mihou.amelia.clients.listeners.impl;

import pw.mihou.amelia.Amelia;
import pw.mihou.amelia.clients.listeners.interfaces.FeedListener;
import pw.mihou.amelia.payloads.AmeliaPayload;
import pw.mihou.amelia.session.AmeliaSession;

public class FeedUpdater implements FeedListener {

    @Override
    public void onEvent(AmeliaPayload payload) {
        Amelia.shards.forEach((integer, api) -> api.getServerTextChannelById(payload.model.getChannel())
                .ifPresent(textChannel -> payload.wrapper.getPubDate()
                        .ifPresent(date -> textChannel.sendMessage(Amelia.format(payload.wrapper, payload.model, textChannel.getServer()))
                                .whenComplete((message, throwable) -> {
                                    AmeliaSession.feedsUpdated.incrementAndGet();

                                    if (throwable != null) {
                                        Amelia.log.error("We were unable to send feed update [{}] to server {}!", payload.model.getUnique(),
                                                textChannel.getServer().getName());
                                    } else {
                                        Amelia.log.info("Amelia has finished (or has attempted) sending feed [{}] to {}!", payload.model.getUnique(),
                                                textChannel.getServer().getName());
                                    }
                                }))));
    }

    @Override
    public String type() {
        return "feed";
    }
}

File: src/main/java/pw/mihou/amelia/clients/listeners/impl/FeedUpdater.kt
package pw.mihou.amelia.clients.listeners.impl

import pw.mihou.amelia.Amelia
import pw.mihou.amelia.io.rome.ItemWrapper
import pw.mihou.amelia.logger
import pw.mihou.amelia.models.FeedModel
import pw.mihou.amelia.nexus
import pw.mihou.amelia.session.AmeliaSession

object FeedUpdater {

    fun onEvent(item: ItemWrapper, feed: FeedModel) {
        nexus.shardManager.getShardOf(feed.server).ifPresent { shard ->
            shard.getServerById(feed.server).flatMap { server -> server.getTextChannelById(feed.channel) }.ifPresent serverCheck@{ channel ->
                if (item.date == null) {
                    logger.error("A feed was received that doesn't contain a date. [feed=${feed.feedUrl}]")
                    return@serverCheck
                }

                channel.sendMessage(Amelia.format(item, feed, channel.server)).thenAccept {
                    AmeliaSession.feedsUpdated.incrementAndGet()
                    logger.info("I have sent a feed update to a server with success. [feed=${feed.feedUrl}, server=${channel.server.id}]")
                }.exceptionally { exception ->
                    logger.error("Failed to send update for a feed to a server. [feed=${feed.feedUrl}, server=${channel.server.id}]", exception)
                    return@exceptionally null
                }
            }
        }
    }

}